File: packages/kotlin/flappy/src/main/kotlin/flappy/llms/Baichuan.kt
package flappy.llms

import com.fasterxml.jackson.annotation.JsonInclude
import flappy.*
import io.ktor.client.*
import io.ktor.client.request.*
import io.ktor.client.statement.*
import io.ktor.http.*
import java.math.BigInteger
import java.security.MessageDigest
import java.time.Instant
import java.util.*

/**
 * https://platform.baichuan-ai.com/docs/api
 */

// https://stackoverflow.com/a/64171625/20030734
internal fun calculateMd5(input: String): String {
  val md = MessageDigest.getInstance("MD5")
  return BigInteger(1, md.digest(input.toByteArray())).toString(16).padStart(32, '0')
}

class Baichuan(
  private val baichuanConfig: BaichuanConfig,
) :
  FlappyLLMBase() {
  private val client = HttpClient()

  override fun close() {
    client.close()
  }

  class BaichuanConfig(
    model: String? = null,
    val apiKey: String, val secretKey: String
  ) {
    val baichuanModel = model ?: "Baichuan2-53B"
  }

  @JsonInclude(JsonInclude.Include.NON_EMPTY)
  data class BaichuanParameters(
    val temperature: Double? = null,
    val top_k: Double? = null,
    val top_p: Double? = null,
    val with_search_enhance: Boolean? = null
  )

  data class BaichuanRequestChatMessage(
    val role: String,
    val content: String
  )

  @JsonInclude(JsonInclude.Include.NON_EMPTY)
  data class BaichuanRequestBody(
    val model: String,
    val messages: List<BaichuanRequestChatMessage>,
    val parameters: BaichuanParameters? = null
  )

  class BaichuanChatMessage(
    val role: String,
    val content: String,
    val finish_reason: String
  )

  class BaichuanResponseData(
    val messages: List<BaichuanChatMessage>
  )

  class BaichuanResponseUsage(
    val prompt_tokens: Int,
    val answer_tokens: Int,
    val total_tokens: Int
  )

  @JsonInclude(JsonInclude.Include.NON_EMPTY)
  class BaichuanResponseBody(
    val code: Int,
    val msg: String,
    val data: BaichuanResponseData?,
    val usage: BaichuanResponseUsage
  )

  override suspend fun chatComplete(
    messages: List<FlappyChatMessage>,
    source: String,
    config: LLMGenerateConfig?
  ): LLMResponse {
    val requestId = UUID.randomUUID().toString()
    val timestamp = Instant.now().epochSecond.toString()

    val body = BaichuanRequestBody(
      model = baichuanConfig.baichuanModel,
      messages = messages.map {
        BaichuanRequestChatMessage(
          role = (when (it.role) {
            FlappyRole.ASSISTANT -> "assistant"
            FlappyRole.SYSTEM -> "system"
            FlappyRole.USER -> "user"
          }),
          content = it.content
        )
      },
      parameters = BaichuanParameters(temperature = config?.temperature, top_p = config?.topP)
    )

    val jsonBody = body.asString()

    val signature = calculateMd5("${baichuanConfig.secretKey}$jsonBody$timestamp")

    logger.info("baichuan input $messages -> $jsonBody, $timestamp")

    try {
      val response = client.post(
        "https://api.baichuan-ai.com/v1/chat"
      ) {
        contentType(ContentType.Application.Json)
        headers {
          append(HttpHeaders.ContentType, "application/json")
          append(HttpHeaders.Authorization, "Bearer ${baichuanConfig.apiKey}")
          append(HttpHeaders.AcceptLanguage, "*")
          append(HttpHeaders.AcceptEncoding, "gzip, deflate")
          append("X-BC-Request-Id", requestId)
          append("X-BC-Timestamp", timestamp)
          append("X-BC-Signature", signature)
          append("X-BC-Sign-Algo", "MD5")
        }
        setBody(jsonBody)
      }

      if (response.status != HttpStatusCode.OK) {
        return LLMResponse.Fail("Invalid status: ${response.status}")
      }

      val respBody = response.bodyAsText()
      logger.info("baichuan output $respBody")

      val result = respBody.castBy(BaichuanResponseBody::class.java)
      if (result.code != 0) return LLMResponse.Fail(result.msg)
      val message = result.data?.messages?.get(0)
      if (message === null) return LLMResponse.Fail("Invalid response")

      return LLMResponse.Success(message.content)

    } catch (e: Exception) {
      return LLMResponse.Fail(e.message ?: "ERROR")
    }
  }
}

File: packages/nodejs/src/llms/baichuan.ts
import { LLMBase } from './llm-base'
import { type ChatMLMessage, type ChatMLResponse, type GenerateConfig } from './interface'
import { randomUUID, createHash } from 'crypto'
import { log } from '../utils'
import { type JsonObject } from 'roarr/dist/types'

// https://platform.baichuan-ai.com/docs/api
type BaichuanModel = string
const BAICHUAN_MODEL: BaichuanModel = 'Baichuan2-53B'

export interface BaichuanArgs {
  baichuan_api_key: string
  baichuan_secret_key: string
  model: BaichuanModel
}

interface BaichuanParameters {
  // [0, 1], default 0.3
  temperature?: number
  // [0, 20], default 5
  top_k?: number
  // [0, 1], default 0.85
  top_p?: number
  // default false
  with_search_enhance?: boolean
}

interface BaichuanRequestBody {
  model: BaichuanModel
  messages: ChatMLMessage[]
  parameters?: BaichuanParameters
}

interface BaichuanChatMLMessage {
  role: 'assistant'
  content: string
  finish_reason: string
}

interface BaichuanResponseUsage {
  prompt_tokens: number
  answer_tokens: number
  total_tokens: number
}

interface BaichuanResponseBody {
  code: number
  msg: string
  data: {
    messages?: BaichuanChatMLMessage[]
  }
  usage: BaichuanResponseUsage
}

const calculateMd5 = (input: string): string => createHash('md5').update(input).digest('hex')

export class Baichuan extends LLMBase {
  private readonly model: BaichuanModel
  private readonly baichuan_api_key: string
  private readonly baichuan_secret_key: string
  override maxTokens = -1

  constructor(args?: BaichuanArgs) {
    super()
    this.model = args?.baichuan_api_key ?? BAICHUAN_MODEL
    this.baichuan_api_key = args?.baichuan_api_key ?? process.env.BAICHUAN_API_KEY!
    if (!this.baichuan_api_key) throw new Error('BAICHUAN_API_KEY not found')
    this.baichuan_secret_key = args?.baichuan_secret_key ?? process.env.BAICHUAN_SECRET_KEY!
    if (!this.baichuan_secret_key) throw new Error('BAICHUAN_SECRET_KEY not found')
  }

  async chatComplete(messages: ChatMLMessage[], config?: GenerateConfig): Promise<ChatMLResponse> {
    const requestId = randomUUID()
    const body: BaichuanRequestBody = {
      model: this.model,
      messages,
      parameters: {
        ...(config?.temperature ? { temperature: config.temperature } : {}),
        ...(config?.top_p ? { top_p: config.top_p } : {})
      }
    }

    const jsonBody = JSON.stringify(body)
    const timestamp = Math.round(new Date().getTime() / 1000).toString()
    const signature = calculateMd5(this.baichuan_secret_key + jsonBody + timestamp)
    const headers = {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${this.baichuan_api_key}`,
      'X-BC-Request-Id': requestId,
      'X-BC-Timestamp': timestamp,
      'X-BC-Signature': signature,
      'X-BC-Sign-Algo': 'MD5'
    }

    try {
      log.debug({ data: { jsonBody, headers } }, 'Request baichuan API')
      const response = await fetch('https://api.baichuan-ai.com/v1/chat', {
        method: 'POST',
        body: jsonBody,
        headers
      })

      if (response.status !== 200) {
        return { success: false, data: `Invalid status: ${response.status}` }
      }

      const result: BaichuanResponseBody = (await response.json()) as BaichuanResponseBody
      log.debug({ data: result } as unknown as JsonObject, 'baichuan API response')

      if (result.code !== 0) return { success: false, data: result.msg }

      const message = result.data?.messages?.[0]
      if (!message) return { success: false, data: 'Invalid response' }

      return { success: true, data: message.content }
    } catch (error) {
      return { success: false, data: undefined }
    }
  }
}
